Make `gears` cross-platform and some other fancy magic

natsukagami opened the pull request:
Most tests on native are running but not all (I ignored `ChannelBehavior` tests but they are still dead ;w;)

- Make gears cross-platform
- Add organization
- Simple fork-join executor
- Add ForkJoin support
- Implement Timer and some efforts to make tests run
- Move scala to 3.3.1
- Get timer to run
- Get other tests to run on all platforms
- Split test framework to support a custom version of munit-native
- Use `FiniteDuration` in schedule
- Make timer tests pass
- Introduce cleanup state for RunnableFuture for... cleanup
- Don't wake up prematurely

natsukagami committed:
Restructure gears into cross-platform with sbt-crossproject

natsukagami committed:
Implement a ForkJoin-executor-based AsyncSupport instance

natsukagami committed:
Implement Timer and some efforts to make tests run

natsukagami committed:
Get other tests to run on all platforms

natsukagami committed:
Document and include build dependencies for Native

natsukagami committed:
Move the timeout test to jvm

m8nmueller commented on shared/src/main/scala/async/Async.scala:
```diff
@@ -47,7 +46,12 @@ object Async:
   def blocking[T](body: Async ?=> T)(using support: AsyncSupport, scheduler: support.Scheduler): T =
-    body(using Blocking(CompletionGroup.Unlinked))
+    val async = Blocking(CompletionGroup.Unlinked)
+    val group = new CompletionGroup
```
~I suggest to just use Async.group~ I will implement this on my thread.

m8nmueller commented on shared/src/main/scala/async/futures.scala:
```diff
@@ -137,7 +141,9 @@ object Future:
             cancellable.listener = listener
             src.onComplete(listener)
-            innerGroup.add(cancellable) // may resume + remove listener immediately
+            if !cleanup then // during cleanup we already cancelled the innerGroup
```
TODO to be removed after my fix

m8nmueller commented on jvm/src/main/scala/async/VThreadSupport.scala:
```diff
@@ -2,19 +2,19 @@ package gears.async
     val th = Thread.startVirtualThread: () =>
-      Thread.sleep(delay)
-      body.run()
-    () => th.interrupt() // TODO this may interrupt the body after sleeping
```
Is there a subject-related reason for removing this note, i.e., did you fix it?

natsukagami replied:
Yeah, I basically just end the sleeping thread and put the `body` into the scheduler; that way the body is run in another v-thread and not get interrupted.

natsukagami committed:
Properly check for completion in future addListener

Otherwise a situation where

Thread A    | Thread B

poll
            | complete
addListener

happens, and A gets deadlocked.

natsukagami committed:
Make sleeper thread a daemon

natsukagami committed:
Use Boehm GC as Immix is crashing sometimes
